Flutter Doctor cannot find Android Studio 4.2

In Android Studio 4.2, the ANDROID_SDK_HOME variable has been deprecated:

*Android Studio 4.2 release notes:*
<img width="517" alt="flutter" src="https://user-images.githubusercontent.com/13650798/117970865-09b0d600-b2de-11eb-920c-158d242d151b.png"> 

  On a clean install of AS and Flutter, `flutter doctor` cannot find Android Studio or JDK.

Initial error:
<img width="455" alt="error" src="https://user-images.githubusercontent.com/13650798/117973083-822b8d00-b306-11eb-9039-e2d9fd77feb5.png">

Temporary fix that (mostly) worked for me:
  1. Manually create JAVA_HOME system variable pointing to Android Studio JRE dir.
  2. Add AS JRE to PATH
  3. Do a `flutter config --android-studio-dir="path\to\studio\not\bin"` 
  *At this point, there will be a thread exception when trying to accept Android Licenses*
  4. Install "Android SDK command line tools" in SDK Manager

Finally, `flutter doctor` claims there are no issues.
<img width="455" alt="success" src="https://user-images.githubusercontent.com/13650798/117974871-a1c3b500-b308-11eb-9003-8a93adb9602c.png">

After installing from zero AS 4.2, flutter doctor found an error: 
 [!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

Also running flutter app on Android failed with this error: 

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:compileDebugKotlin'.
> Kotlin could not find the required JDK tools in the Java installation 'C:\Program Files\Java\jre1.8.0_291' used by Gradle. Make sure Gradle is running on a JDK, not JRE.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 7s
Exception: Gradle task assembleDebug failed with exit code 1


Has anyone of you encountered with this issue @JCKodel  @robjvan?

--- a/packages/flutter_tools/lib/src/android/android_studio.dart
+++ b/packages/flutter_tools/lib/src/android/android_studio.dart
@@ -397,6 +397,30 @@ class AndroidStudio implements Comparable<AndroidStudio> {
       }
     }
 
+    // 4.2 has a different location for AndroidStudio installs on Windows.
+    if (globals.platform.isWindows && globals.platform.environment.containsKey('LOCALAPPDATA')) {
+      final File homeDot = globals.fs.file(globals.fs.path.join(
+        globals.platform.environment['LOCALAPPDATA']!,
+        'Google',
+        'AndroidStudio4.2',
+        '.home',
+      ));
+      if (homeDot.existsSync()) {
+        final String installPath = homeDot.readAsStringSync();
+        if (globals.fs.isDirectorySync(installPath)) {
+          final AndroidStudio studio = AndroidStudio(
+            installPath,
+            version: Version(4, 2, 0),
+            studioAppName: 'Android Studio 4.2',
+          );
+          if (studio != null && !_hasStudioAt(studio.directory, newerThan: studio.version)) {
+            studios.removeWhere((AndroidStudio other) => other.directory == studio.directory);
+            studios.add(studio);
+          }
+        }
+      }
+    }
+
     final String? configuredStudioDir = globals.config.getValue('android-studio-dir') as String?;
     if (configuredStudioDir != null && !_hasStudioAt(configuredStudioDir)) {
       studios.add(AndroidStudio(configuredStudioDir,

--- a/packages/flutter_tools/test/general.shard/android/android_studio_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/android_studio_test.dart
@@ -408,9 +408,45 @@ void main() {
     FileSystem: () => windowsFileSystem,
     ProcessManager: () => FakeProcessManager.any(),
   });
+  
+  testUsingContext('Can discover Android Studio 4.2 location on Windows', () {
+    windowsFileSystem.file(r'C:\Users\Dash\AppData\Local\Google\AndroidStudio4.2\.home')
+      ..createSync(recursive: true)
+      ..writeAsStringSync(r'C:\Program Files\AndroidStudio');
+    windowsFileSystem
+      .directory(r'C:\Program Files\AndroidStudio')
+      .createSync(recursive: true);
+
+    final AndroidStudio studio = AndroidStudio.allInstalled().single;
+
+    expect(studio.version, Version(4, 2, 0));
+    expect(studio.studioAppName, 'Android Studio 4.2');
+  }, overrides: <Type, Generator>{
+    Platform: () => windowsPlatform,
+    FileSystem: () => windowsFileSystem,
+    ProcessManager: () => FakeProcessManager.any(),
+  });
 
   testUsingContext('Does not discover Android Studio 4.1 location on Windows if LOCALAPPDATA is null', () {
     windowsFileSystem.file(r'C:\Users\Dash\AppData\Local\Google\AndroidStudio4.1\.home')
+      ..createSync(recursive: true)
+      ..writeAsStringSync(r'C:\Program Files\AndroidStudio');
+    windowsFileSystem
+      .directory(r'C:\Program Files\AndroidStudio')
+      .createSync(recursive: true);
+
+    expect(AndroidStudio.allInstalled(), isEmpty);
+  }, overrides: <Type, Generator>{
+    Platform: () => FakePlatform(
+      operatingSystem: 'windows',
+      environment: <String, String>{}, // Does not include LOCALAPPDATA
+    ),
+    FileSystem: () => windowsFileSystem,
+    ProcessManager: () => FakeProcessManager.any(),
+  });
+
+  testUsingContext('Does not discover Android Studio 4.2 location on Windows if LOCALAPPDATA is null', () {
+    windowsFileSystem.file(r'C:\Users\Dash\AppData\Local\Google\AndroidStudio4.2\.home')
       ..createSync(recursive: true)
       ..writeAsStringSync(r'C:\Program Files\AndroidStudio');
     windowsFileSystem

--- a/packages/flutter_tools/test/general.shard/android/android_studio_test.dart
+++ b/packages/flutter_tools/test/general.shard/android/android_studio_test.dart
@@ -408,7 +408,7 @@ void main() {
     FileSystem: () => windowsFileSystem,
     ProcessManager: () => FakeProcessManager.any(),
   });
-  
+
   testUsingContext('Can discover Android Studio 4.2 location on Windows', () {
     windowsFileSystem.file(r'C:\Users\Dash\AppData\Local\Google\AndroidStudio4.2\.home')
       ..createSync(recursive: true)

